diff --git a/Adafruit_GFX.cpp b/Adafruit_GFX.cpp
index 0631d43..1673541 100644
--- a/Adafruit_GFX.cpp
+++ b/Adafruit_GFX.cpp
@@ -1465,7 +1465,8 @@ void Adafruit_GFX::getTextBounds(const char *str, int16_t x, int16_t y,
 
   if (maxx >= minx) {     // If legit string bounds were found...
     *x1 = minx;           // Update x1 to least X coord,
-    *w = maxx - minx + 1; // And w to bound rect width
+    //*w = maxx - minx + 1; // And w to bound rect width
+    *w = maxx + 1;
   }
   if (maxy >= miny) { // Same for height
     *y1 = miny;
