diff --git a/teensy4/AudioStream.cpp b/teensy4/AudioStream.cpp
index 232928b..d7f6dd3 100644
--- a/teensy4/AudioStream.cpp
+++ b/teensy4/AudioStream.cpp
@@ -52,7 +52,7 @@ void software_isr(void);
 
 // Set up the pool of audio data blocks
 // placing them all onto the free list
-FLASHMEM void AudioStream::initialize_memory(audio_block_t *data, unsigned int num)
+FLASHMEM void AudioStream::initialize_memory(audio_block_t *data, unsigned int num, int16_t *dataBuffers)
 {
 	unsigned int i;
 	unsigned int maxnum = MAX_AUDIO_MEMORY / AUDIO_BLOCK_SAMPLES / 2;
@@ -71,6 +71,9 @@ FLASHMEM void AudioStream::initialize_memory(audio_block_t *data, unsigned int n
 	}
 	for (i=0; i < num; i++) {
 		data[i].memory_pool_index = i;
+		if (dataBuffers) {
+			data[i].data = dataBuffers + i*AUDIO_BLOCK_SAMPLES;
+		}
 	}
 	__enable_irq();
 
@@ -183,7 +186,7 @@ audio_block_t * AudioStream::receiveWritable(unsigned int index)
 	inputQueue[index] = NULL;
 	if (in && in->ref_count > 1) {
 		p = allocate();
-		if (p) memcpy(p->data, in->data, sizeof(p->data));
+		if (p) memcpy(p->data, in->data, sizeof(int16_t) * AUDIO_BLOCK_SAMPLES);
 		in->ref_count--;
 		in = p;
 	}
