diff --git a/teensy4/imxrt.h b/teensy4/imxrt.h
index cb5cafe..def7fd8 100644
--- a/teensy4/imxrt.h
+++ b/teensy4/imxrt.h
@@ -9993,4 +9993,9 @@ struct arm_fault_info_struct {
 	uint32_t crc;  // crc must be last
 };
 
-
+// Breadcrumbs stored in the top 128 bytes of OCRAM (at 0x2027FFC0)
+struct crashreport_breadcrumbs_struct {
+	uint32_t bitmask;
+	uint32_t value[6];
+	uint32_t checksum; // currently unused
+};
