diff --git a/teensy4/Print.h b/teensy4/Print.h
index b3d6006..cfa74af 100644
--- a/teensy4/Print.h
+++ b/teensy4/Print.h
@@ -114,6 +114,7 @@ class Print
 	// https://forum.pjrc.com/threads/62473?p=256873&viewfull=1#post256873
 	int printf(const char *format, ...) /*__attribute__ ((format (printf, 2, 3)))*/;
 	int printf(const __FlashStringHelper *format, ...);
+    int vprintf(const char * format, va_list arg);
   protected:
 	void setWriteError(int err = 1) { write_error = err; }
   private:
