diff --git a/teensy4/usb.c b/teensy4/usb.c
index f7f3156..5687767 100644
--- a/teensy4/usb.c
+++ b/teensy4/usb.c
@@ -16,6 +16,7 @@
 #include "avr/pgmspace.h"
 #include <string.h>
 #include "debug/printf.h"
+#include "AudioSampleRate.h"
 
 //#define LOG_SIZE  20
 //uint32_t transfer_log_head=0;
@@ -128,7 +129,7 @@ FLASHMEM void usb_init(void)
 	// assume PLL3 is already running - already done by usb_pll_start() in main.c
 
 	CCM_CCGR6 |= CCM_CCGR6_USBOH3(CCM_CCGR_ON); // turn on clocks to USB peripheral
-	
+
 	printf("BURSTSIZE=%08lX\n", USB1_BURSTSIZE);
 	//USB1_BURSTSIZE = USB_BURSTSIZE_TXPBURST(4) | USB_BURSTSIZE_RXPBURST(4);
 	USB1_BURSTSIZE = 0x0404;
@@ -661,8 +662,8 @@ static void endpoint0_setup(uint64_t setupdata)
 		break;
 	  case 0x81A2: // GET_CUR (wValue=0, wIndex=interface, wLength=len)
 		if (setup.wLength >= 3) {
-			endpoint0_buffer[0] = 44100 & 255;
-			endpoint0_buffer[1] = 44100 >> 8;
+			endpoint0_buffer[0] = (unsigned)(AUDIO_SAMPLE_RATE_HZ_F) & 255;
+			endpoint0_buffer[1] = (unsigned)(AUDIO_SAMPLE_RATE_HZ_F) >> 8;
 			endpoint0_buffer[2] = 0;
 			endpoint0_transmit(endpoint0_buffer, 3, 0);
 			return;
