diff --git a/teensy4/usb_audio.h b/teensy4/usb_audio.h
index a53d6be..95c0f97 100644
--- a/teensy4/usb_audio.h
+++ b/teensy4/usb_audio.h
@@ -31,6 +31,7 @@
 #pragma once
 
 #include "usb_desc.h"
+#include "usb_dev.h"
 #ifdef AUDIO_INTERFACE
 
 #define FEATURE_MAX_VOLUME 0xFF  // volume accepted from 0 to 0xFF
@@ -44,10 +45,25 @@ extern uint16_t usb_audio_transmit_buffer[];
 extern uint32_t usb_audio_sync_feedback;
 extern uint8_t usb_audio_receive_setting;
 extern uint8_t usb_audio_transmit_setting;
+extern uint8_t usb_audio_sync_nbytes;
+extern uint8_t usb_audio_sync_rshift;
 extern void usb_audio_receive_callback(unsigned int len);
 extern unsigned int usb_audio_transmit_callback(void);
 extern int usb_audio_set_feature(void *stp, uint8_t *buf);
 extern int usb_audio_get_feature(void *stp, uint8_t *data, uint32_t *datalen);
+extern uint32_t feedback_accumulator;
+extern volatile uint32_t usb_audio_underrun_count;
+extern volatile uint32_t usb_audio_overrun_count;
+
+extern transfer_t rx_transfer;
+extern transfer_t sync_transfer;
+extern transfer_t tx_transfer;
+
+extern void rx_event(transfer_t *t);
+extern void sync_event(transfer_t *t);
+extern void copy_to_buffers(const uint32_t *src, int16_t *left, int16_t *right, unsigned int len);
+extern void tx_event(transfer_t *t);
+extern void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsigned int len);
 #ifdef __cplusplus
 }
 #endif
@@ -62,6 +78,11 @@ struct usb_audio_features_struct {
 #ifdef __cplusplus
 #include "AudioStream.h"
 
+namespace SysPlatform { // forward declares
+class SysAudioInputUsb;
+class SysAudioOutputUsb;
+}
+
 class AudioInputUSB : public AudioStream
 {
 public:
@@ -72,7 +93,7 @@ public:
 	friend int usb_audio_set_feature(void *stp, uint8_t *buf);
 	friend int usb_audio_get_feature(void *stp, uint8_t *data, uint32_t *datalen);
 	static struct usb_audio_features_struct features;
-	float volume(void) {
+	static float volume(void) {
 		if (features.mute) return 0.0;
 		return (float)(features.volume) * (1.0 / (float)FEATURE_MAX_VOLUME);
 	}
@@ -84,6 +105,7 @@ private:
 	static audio_block_t *ready_right;
 	static uint16_t incoming_count;
 	static uint8_t receive_flag;
+    friend SysPlatform::SysAudioInputUsb;
 };
 
 class AudioOutputUSB : public AudioStream
@@ -101,6 +123,7 @@ private:
 	static audio_block_t *right_2nd;
 	static uint16_t offset_1st;
 	audio_block_t *inputQueueArray[2];
+    friend SysPlatform::SysAudioOutputUsb;
 };
 #endif // __cplusplus
 
