diff --git a/teensy4/usb_audio.cpp b/teensy4/usb_audio.cpp
index fa492a5..603d287 100644
--- a/teensy4/usb_audio.cpp
+++ b/teensy4/usb_audio.cpp
@@ -67,7 +67,7 @@ volatile uint32_t usb_audio_underrun_count;
 volatile uint32_t usb_audio_overrun_count;
 
 
-static void rx_event(transfer_t *t)
+void rx_event(transfer_t *t)
 {
 	if (t) {
 		int len = AUDIO_RX_SIZE - ((rx_transfer.status >> 16) & 0x7FFF);
@@ -79,7 +79,7 @@ static void rx_event(transfer_t *t)
 	usb_receive(AUDIO_RX_ENDPOINT, &rx_transfer);
 }
 
-static void sync_event(transfer_t *t)
+void sync_event(transfer_t *t)
 {
 	// USB 2.0 Specification, 5.12.4.2 Feedback, pages 73-75
 	//printf("sync %x\n", sync_transfer.status); // too slow, can't print this much
@@ -94,7 +94,8 @@ void usb_audio_configure(void)
 	printf("usb_audio_configure\n");
 	usb_audio_underrun_count = 0;
 	usb_audio_overrun_count = 0;
-	feedback_accumulator = 739875226; // 44.1 * 2^24
+	//feedback_accumulator = 739875226; // 44.1 * 2^24
+    feedback_accumulator = (AUDIO_SAMPLE_RATE_HZ_F / 1000.0f) * 16777216.0f;  // Sample rate in Khz * 2^24
 	if (usb_high_speed) {
 		usb_audio_sync_nbytes = 4;
 		usb_audio_sync_rshift = 8;
@@ -129,9 +130,9 @@ void AudioInputUSB::begin(void)
 	update_responsibility = false;
 }
 
-static void copy_to_buffers(const uint32_t *src, int16_t *left, int16_t *right, unsigned int len)
+void copy_to_buffers(const uint32_t *src, int16_t *left, int16_t *right, unsigned int len)
 {
-	uint32_t *target = (uint32_t*) src + len; 
+	uint32_t *target = (uint32_t*) src + len;
 	while ((src < target) && (((uintptr_t) left & 0x02) != 0)) {
 		uint32_t n = *src++;
 		*left++ = n & 0xFFFF;
@@ -297,7 +298,7 @@ uint16_t AudioOutputUSB::offset_1st;
 /*DMAMEM*/ uint16_t usb_audio_transmit_buffer[AUDIO_TX_SIZE/2] __attribute__ ((used, aligned(32)));
 
 
-static void tx_event(transfer_t *t)
+void tx_event(transfer_t *t)
 {
 	int len = usb_audio_transmit_callback();
 	usb_audio_sync_feedback = feedback_accumulator >> usb_audio_sync_rshift;
@@ -314,7 +315,7 @@ void AudioOutputUSB::begin(void)
 	right_1st = NULL;
 }
 
-static void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsigned int len)
+void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsigned int len)
 {
 	// TODO: optimize...
 	while (len > 0) {
@@ -323,15 +324,17 @@ static void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsi
 	}
 }
 
+
 void AudioOutputUSB::update(void)
 {
 	audio_block_t *left, *right;
 
 	// TODO: we shouldn't be writing to these......
-	//left = receiveReadOnly(0); // input 0 = left channel
-	//right = receiveReadOnly(1); // input 1 = right channel
-	left = receiveWritable(0); // input 0 = left channel
-	right = receiveWritable(1); // input 1 = right channel
+	left = receiveReadOnly(0); // input 0 = left channel
+	right = receiveReadOnly(1); // input 1 = right channel
+	//left = receiveWritable(0); // input 0 = left channel
+	//right = receiveWritable(1); // input 1 = right channel
+
 	if (usb_audio_transmit_setting == 0) {
 		if (left) release(left);
 		if (right) release(right);
@@ -348,15 +351,17 @@ void AudioOutputUSB::update(void)
 			if (right) release(right);
 			return;
 		}
-		memset(left->data, 0, sizeof(left->data));
+		//memset(left->data, 0, sizeof(left->data));
+		memset(left->data, 0, sizeof(int16_t)*AUDIO_SAMPLES_PER_BLOCK);
 	}
 	if (right == NULL) {
 		right = allocate();
 		if (right == NULL) {
-			release(left);
+			if (left) release(left);
 			return;
 		}
-		memset(right->data, 0, sizeof(right->data));
+		//memset(right->data, 0, sizeof(right->data));
+		memset(right->data, 0, sizeof(int16_t)*AUDIO_SAMPLES_PER_BLOCK);
 	}
 	__disable_irq();
 	if (left_1st == NULL) {
@@ -389,16 +394,16 @@ void AudioOutputUSB::update(void)
 // no data to transmit
 unsigned int usb_audio_transmit_callback(void)
 {
-	static uint32_t count=5;
-	uint32_t avail, num, target, offset, len=0;
+	//static uint32_t count=5;
+	uint32_t avail, num, target =48, offset, len=0;
 	audio_block_t *left, *right;
 
-	if (++count < 10) {   // TODO: dynamic adjust to match USB rate
-		target = 44;
-	} else {
-		count = 0;
-		target = 45;
-	}
+	// if (++count < 10) {   // TODO: dynamic adjust to match USB rate
+	// 	target = 48;
+	// } else {
+	// 	count = 0;
+	// 	target = 49;
+	// }
 	while (len < target) {
 		num = target - len;
 		left = AudioOutputUSB::left_1st;
@@ -498,7 +503,7 @@ int usb_audio_get_feature(void *stp, uint8_t *data, uint32_t *datalen)
 	return 0;
 }
 
-int usb_audio_set_feature(void *stp, uint8_t *buf) 
+int usb_audio_set_feature(void *stp, uint8_t *buf)
 {
 	struct setup_struct setup = *((struct setup_struct *)stp);
 	if (setup.bmRequestType==0x21) { // should check bRequest, bChannel and UnitID
