diff --git a/teensy4/usb_midi.c b/teensy4/usb_midi.c
index 77a7ac7..d3e6680 100644
--- a/teensy4/usb_midi.c
+++ b/teensy4/usb_midi.c
@@ -162,7 +162,9 @@ void usb_midi_write_packed(uint32_t n)
 		if (!usb_configuration) return;
 		yield();
 	}
-	uint32_t *txdata = (uint32_t *)(txbuffer + (tx_head * TX_SIZE) + (TX_SIZE - tx_available));
+	//uint32_t *txdata = (uint32_t *)(txbuffer + (tx_head * TX_SIZE) + (TX_SIZE - tx_available));
+	// https://github.com/PaulStoffregen/cores/commit/70ba01accd728abe75ebfc8dcd8b3d3a8f3e3f25
+	uint32_t *txdata = (uint32_t *)(txbuffer + (tx_head * TX_SIZE) + (tx_packet_size - tx_available));
 	*txdata = n;
 	tx_available -= 4;
 	if (tx_available == 0) {
